packages/deployments/contracts/contracts_forge/core/xreceivers/Unwrapper.t.sol

```diff
@@ -0,0 +1,239 @@
+      initialRecipientBalance + amount,
+      "recipient didn't receive ETH! did we record initial balances?"
+    );
+
+    require(
+      initialUnwrapperBalance >= amount,
+      "initial balance of unwrapper was less than amount... did we record initial balances?"
+    );
+
+    // Unwrapper should have been initially funded `amount` in ETH, and should have been debited all their ETH in most cases.
+    assertEq(
+      address(unwrapper).balance,
+      initialUnwrapperBalance - amount,
+      "unwrapper didn't send ETH! did we record initial balances?"
+    );
+  }
+
+  // ============ Unwrapper.xReceive ============
+  function test_Unwrapper__xReceive_works(uint256 amount) public {
+    if (amount == 0) {
+      return; // Skip 0 amount case.
+    }
+
+    address asset = MOCK_WRAPPER;
+    bytes memory callData = abi.encode(recipient);
+
+    utils_expectRecipientToReceive(recipient, asset, amount);
+    utils_recordInitialBalances(recipient);
+
+    vm.prank(MOCK_CONNEXT);
+    unwrapper.xReceive(transferId, amount, asset, originSender, origin, callData);
+
+    utils_assertSentEthToRecipient(recipient, amount);
+  }
+
+  function test_Unwrapper__xReceive_worksIfRecipientIsZero() public {
+    address asset = MOCK_WRAPPER;
+    uint256 amount = 10 ether;
+
+    utils_expectRecipientToReceive(address(0), asset, amount);
+    utils_recordInitialBalances(address(0));
+
+    vm.prank(MOCK_CONNEXT);
+    unwrapper.xReceive(transferId, amount, asset, originSender, origin, callDataNoRecipient);
+
+    utils_assertSentEthToRecipient(address(0), amount);
+  }
+
+  function test_Unwrapper__xReceive_failsIfNotConnext() public {
+    address asset = MOCK_WRAPPER;
+    uint256 amount = 10 ether;
+
+    vm.expectRevert("unwrap: !connext");
+    unwrapper.xReceive(transferId, amount, asset, originSender, origin, callDataWithRecipient);
+  }
+
+  function test_Unwrapper__xReceive_failsIfZeroAmount() public {
+    address asset = MOCK_WRAPPER;
+    uint256 amount = 0;
+
+    vm.expectRevert("unwrap: !amount");
+    vm.prank(MOCK_CONNEXT);
+    unwrapper.xReceive(transferId, amount, asset, originSender, origin, callDataWithRecipient);
+  }
+
+  function test_Unwrapper__xReceive_transfersIfNonWrapperAsset() public {
+    address asset = MOCK_ERC20; // xReceive will be getting a random non-wrapper erc20 asset.
+    uint256 amount = 10 ether;
+
+    vm.expectEmit(true, true, true, true);
+    emit WrongAsset(recipient, asset);
+
+    vm.clearMockedCalls();
+    // Even though the transfer returns `false`, indicating it failed, we should still procceed
+    // as if it did not. Assets will always be sweepable at a later point.
+    utils_setUpMockErc20(false);
+
+    utils_expectRecipientToReceive(recipient, asset, amount);
+
+    vm.prank(MOCK_CONNEXT);
+    unwrapper.xReceive(transferId, amount, asset, originSender, origin, callDataWithRecipient);
+  }
+
+  function test_Unwrapper__xReceive_emitsEventIfNonWrapperAssetAndERC20TransferFails() public {
```

Comment:
we should have some assertions the asset remained on the wrapper and the recipient didn't get anything

Reply:
erm wait this test isn't testing what it should, it should test that the function will revert